stan/math/prim/fun/binomial_coefficient_log.hpp

```diff
@@ -48,32 +57,91 @@ namespace math {
    \f[
    \frac{\partial\, \mbox{binomial\_coefficient\_log}(x, y)}{\partial y} =
    \begin{cases}
-     \textrm{error} & \mbox{if } y > x \textrm{ or } y < 0\\
+     \textrm{error} & \mbox{if } y > x + 1 \textrm{ or } y < -1 \textrm{ or } x
+ < -1\\
      -\Psi(y+1) & \mbox{if } 0\leq y \leq x \\
      \quad +\Psi(x-y+1)& \\[6pt]
      \textrm{NaN} & \mbox{if } x = \textrm{NaN or } y = \textrm{NaN}
    \end{cases}
    \f]
  *
- * @tparam T_N type of the first argument
- * @tparam T_n type of the second argument
- * @param N total number of objects.
- * @param n number of objects chosen.
- * @return log (N choose n).
+ *  This function is numerically more stable than naive evaluation via lgamma.
+ *
+ * @tparam T_n type of the first argument
+ * @tparam T_k type of the second argument
+ *
+ * @param n total number of objects.
+ * @param k number of objects chosen.
+ * @return log (n choose k).
  */
-template <typename T_N, typename T_n>
-inline return_type_t<T_N, T_n> binomial_coefficient_log(const T_N N,
-                                                        const T_n n) {
-  const double CUTOFF = 1000;
-  if (N - n < CUTOFF) {
-    const T_N N_plus_1 = N + 1;
-    return lgamma(N_plus_1) - lgamma(n + 1) - lgamma(N_plus_1 - n);
+
+template <typename T_n, typename T_k>
+inline return_type_t<T_n, T_k> binomial_coefficient_log(const T_n n,
+                                                        const T_k k) {
+  if (is_any_nan(n, k)) {
+    return stan::math::NOT_A_NUMBER;
+  }
+
+  // Choosing the more stable of the symmetric branches
+  if (n > 0 && k > value_of_rec(n) / 2.0 + 1e-8) {
+    return binomial_coefficient_log(n, n - k);
+  }
+
+  using T_partials_return = partials_return_t<T_n, T_k>;
```

Comment:
We are moving to having the `using` statements to be at the start of each function.